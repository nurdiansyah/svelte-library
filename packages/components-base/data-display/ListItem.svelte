<script lang="ts">
  import type { SvelteComponent } from "svelte/internal";
  import type { AvatarProps, ListItemProps } from "./types";

  import { clsx } from "../utils";
  import { getContext } from "./list/context";
  import Icon from "./Icon.svelte";
  import Div from "../utils/Div.svelte";
  import Button from "../inputs/Button.svelte";
  import Avatar from "@deboxsoft/svelte-components-base/data-display/Avatar.svelte";

  let { class: className } = $$props;
  let classes: string | undefined;
  const ctx = getContext();

  export let alignItems: ListItemProps["alignItems"] = "center";
  export let autofocus: ListItemProps["autofocus"] = false;
  export let avatar: ListItemProps["avatar"] = undefined;
  export let button: ListItemProps["button"] = false;
  export let component: ListItemProps["component"] = undefined;
  export let dense: ListItemProps["dense"] = false;
  export let disabled: ListItemProps["disabled"] = false;
  export let disableGutter: ListItemProps["disableGutter"] = false;
  export let divider: ListItemProps["divider"] = false;
  export let href: ListItemProps["href"] = undefined;
  export let icon: ListItemProps["icon"] = undefined;
  export let iconProps: ListItemProps["iconProps"] = {};
  export let label: ListItemProps["label"] = undefined;
  export let labelInset: ListItemProps["labelInset"] = false;
  export let secondaryAction: ListItemProps["secondaryAction"] = undefined;
  export let selected: ListItemProps["selected"] = false;
  export let subLabel: ListItemProps["subLabel"] = undefined;

  let node: HTMLElement, Component: typeof SvelteComponent, avatarProps: Partial<AvatarProps>;

  $: Component = component || (href || button ? Button : Div);
  $: {
    classes = clsx(
      className,
      "list-item",
      (dense || ctx.dense) && "-dense",
      !disableGutter && "-gutters",
      divider && "-divider",
      disabled && "-disabled",
      button && "-button",
      alignItems === "flex-start" && "-align-items-flex-start",
      $$slots["secondary-action"] || (secondaryAction && "-has-secondary-action"),
      selected && "-selected"
    );
  }

  $: avatarProps = avatar && (typeof avatar === "string" ? { src: avatar } : avatarProps);

  $: {
    if (node && node.focus && autofocus) {
      node.focus();
    }
  }
</script>

<svelte:component this={Component} {...$$restProps} class={classes} {href} {node}>
  {#if $$slots['avatar'] || avatarProps}
    <slot name="avatar" classAvatar="avatar">
      <Avatar {...avatarProps} class="avatar" />
    </slot>
  {/if}
  <slot name="icon" classIcon="icon">
    {#if icon}
      <Icon class="icon" component={icon} {...iconProps} />
    {/if}
  </slot>
  <div class="item-label" class:-multiline={$$slots['sub-label']} class:-inset={labelInset}>
    <div class="label">
      <slot>{label}</slot>
    </div>
    {#if $$slots['sub-label'] || subLabel}
      <slot name="sub-label" classSubLabel="sub-label">
        <div class="sub-label">{subLabel}</div>
      </slot>
    {/if}
  </div>
  {#if $$slots['secondary-action']}
    <div class="secondary-action">
      <slot name="secondary-action" />
    </div>
  {/if}
</svelte:component>
