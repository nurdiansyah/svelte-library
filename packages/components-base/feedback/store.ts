import { writable } from "svelte/store";
import type { Optional } from "utility-types";

interface Notifier {
  message: string;
  type: string;
  subMessage?: string;
  icon?: string;
  action?: string;
  unread?: boolean;
}

type Toast = Notifier & { timeout: number };
type NotifyOptions = Optional<Toast, "timeout">;
type ToastOptions = Omit<NotifyOptions, "type">;
type ToastStore = Toast[];
type NotifierStore = Notifier[];

const sendToast = (options: NotifyOptions) => {
  options.timeout = options.timeout || 10;
  toastStore.update((value) => [...value, options as Toast]);
};

const sendNotify = (options: NotifyOptions) => {
  if (options.timeout) {
    sendToast(options);
    delete options.timeout;
  }
  notifierStore.update((value) => [...value, options as Notifier]);
};

const toastStore = writable<ToastStore>([]);

const notifierStore = writable<NotifierStore>([]);

export const toast = {
  info: (options: ToastOptions) => sendToast({ type: "info", ...options }),
  success: (options: ToastOptions) => sendToast({ type: "success", ...options }),
  warning: (options: ToastOptions) => sendToast({ type: "warning", ...options }),
  error: (options: ToastOptions) => sendToast({ type: "error", ...options }),
  danger: (options: ToastOptions) => sendToast({ type: "danger", ...options }),
  reset: () => toastStore.set([]),
  subscribe: toastStore.subscribe,
  store: notifierStore
};

export const notify = {
  info: (options: ToastOptions) => sendNotify({ type: "info", ...options }),
  success: (options: ToastOptions) => sendNotify({ type: "success", ...options }),
  warning: (options: ToastOptions) => sendNotify({ type: "warning", ...options }),
  error: (options: ToastOptions) => sendNotify({ type: "error", ...options }),
  danger: (options: ToastOptions) => sendNotify({ type: "danger", ...options }),
  reset: () => notifierStore.set([]),
  subscribe: notifierStore.subscribe,
  store: notifierStore
};
