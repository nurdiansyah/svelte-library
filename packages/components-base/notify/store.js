"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.notify = exports.toast = void 0;
const store_1 = require("svelte/store");
const sendToast = (options) => {
    options.timeout = options.timeout || 10;
    toastStore.update((value) => [...value, options]);
};
const sendNotify = (options) => {
    if (options.timeout) {
        sendToast(options);
        delete options.timeout;
    }
    notifierStore.update((value) => [...value, options]);
};
const toastStore = store_1.writable([]);
const notifierStore = store_1.writable([]);
exports.toast = {
    info: (options) => sendToast({ type: "info", ...options }),
    success: (options) => sendToast({ type: "success", ...options }),
    warning: (options) => sendToast({ type: "warning", ...options }),
    error: (options) => sendToast({ type: "error", ...options }),
    danger: (options) => sendToast({ type: "danger", ...options }),
    reset: () => toastStore.set([]),
    subscribe: toastStore.subscribe,
    store: notifierStore
};
exports.notify = {
    info: (options) => sendNotify({ type: "info", ...options }),
    success: (options) => sendNotify({ type: "success", ...options }),
    warning: (options) => sendNotify({ type: "warning", ...options }),
    error: (options) => sendNotify({ type: "error", ...options }),
    danger: (options) => sendNotify({ type: "danger", ...options }),
    reset: () => notifierStore.set([]),
    subscribe: notifierStore.subscribe,
    store: notifierStore
};
