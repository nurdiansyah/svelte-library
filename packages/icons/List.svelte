<script>
  export let disabled;
  export let label;
  export let titleAccess;
  export let viewBox = "0 0 24 24";
  export let htmlColor;

</script>

<svg
  {...$$restProps}
  viewBox="{viewBox}"
  aria-hidden="{titleAccess ? undefined : 'true'}"
  color="{htmlColor}"
  role="{titleAccess ? 'img' : undefined}"
>
    <path d="M3 13h2v-2H3v2zm0 4h2v-2H3v2zm0-8h2V7H3v2zm4 4h14v-2H7v2zm0 4h14v-2H7v2zM7 7v2h14V7H7z" />
    {#if titleAccess}
      <title>{titleAccess}</title>
    {/if}
</svg>

