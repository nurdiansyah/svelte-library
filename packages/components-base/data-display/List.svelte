<script lang="ts">
  import type { SvelteComponent } from "svelte/internal";
  import type { DataList } from "./types";

  import { clsx } from "../utils";
  import { setContext } from "./list/context";
  import ListItem from "./ListItem.svelte";
  import Icon from "@deboxsoft/svelte-components-base/data-display/Icon.svelte";

  // state
  let { class: className } = $$props;
  let classes: string | undefined;

  // props
  export let disablePadding: boolean = false;
  export let dataList: DataList = [];
  export let dense: boolean = false;
  export let subheader: SvelteComponent | undefined = undefined;
  export let node: HTMLElement | undefined = undefined;

  // context
  setContext({ dense });

  // handler

  // reactive
  $: classes = clsx(
    className,
    "dbx-list",
    ($$slots.subheader || subheader) && "-subheader",
    !disablePadding && "-padding",
    dense && "-dense"
  );
</script>

<div class={classes} {...$$restProps} role="menu" on:keydown bind:this={node}>
  {#if $$slots['subheader']}
    <div class="subheader">
      <slot name="subheader" />
    </div>
  {/if}
  {#each dataList as data}
    <div class="item">
      <slot {data}>
        <ListItem {...data}>{data.label}</ListItem>
      </slot>
    </div>
    {#if data.dataList}
      <svelte:self {...$$props} dataList={data.dataList} />
    {/if}
  {/each}
</div>

<style src="./styles/list.scss" global>
</style>
