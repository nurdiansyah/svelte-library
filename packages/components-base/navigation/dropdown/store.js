import { writable, derived } from "svelte/store";
export const dropdownStore = writable({});
export const setActive = () => dropdownStore.update((value) => {
    value.active = true;
    return value;
});
export const setItemActive = (active) => dropdownStore.update((value) => {
    value.itemActive = active;
    return value;
});
export const isActive = derived(dropdownStore, (values) => values.active);
export const isOpen = derived(dropdownStore, (values) => values.open);
