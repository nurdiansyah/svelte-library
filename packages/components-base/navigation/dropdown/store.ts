import { writable, derived } from "svelte/store";

type DropdownStore = {
  active?: boolean;
  open?: boolean;
  itemActive?: boolean;
  direction?: "left" | "up" | "right";
  toggle?: Function;
};

export const dropdownStore = writable<DropdownStore>({});

export const setActive = () =>
  dropdownStore.update((value) => {
    value.active = true;
    return value;
  });

export const setItemActive = (active: boolean) =>
  dropdownStore.update((value) => {
    value.itemActive = active;
    return value;
  });

export const isActive = derived(dropdownStore, (values) => values.active);
export const isOpen = derived(dropdownStore, (values) => values.open);
