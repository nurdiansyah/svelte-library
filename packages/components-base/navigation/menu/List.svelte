<script lang="ts">
  import { onMount } from "svelte";
  import List from "../../data-display/List.svelte";
  import { clsx } from "../../utils";
  import { ownerDocument } from "../../utils/dom";
  import type { DataList } from "../../types";

  // state
  let { class: className } = $$props;
  let listNode: HTMLElement;
  let textCriteria: {
    keys: [];
    repeating: true;
    previousKeyMatched: true;
    lastTime: null;
  };
  let classes: string | undefined;

  // props
  export let autoFocus: boolean = false;
  export let autoFocusItem: boolean = false;
  export let disableItemsFocusable: boolean = false;
  export let disableListWrap: boolean = false;
  export let onKeyDown: ((e?: Event) => void) | undefined = undefined;
  export let variant: "menu" | "selectedMenu";
  export let dataList: DataList;

  onMount(() => {
    init();
  });

  function init() {
    if (listNode) {
    }
  }

  function nextItem(nodeFocus?: HTMLElement, _disableListWrap?: boolean) {
    if (nodeFocus === listNode) {
      return listNode.firstChild;
    }
    // nodeFocus is Item list
    if (nodeFocus && nodeFocus.nextElementSibling) {
      return nodeFocus.nextElementSibling;
    }

    // tidak node focus
    return _disableListWrap ? null : listNode.firstChild;
  }

  function previousItem(nodeFocus?: HTMLElement, _disableListWrap?: boolean) {
    if (nodeFocus === listNode) {
      return _disableListWrap ? listNode.firstChild : listNode.lastChild;
    }

    if (nodeFocus && nodeFocus.previousElementSibling) {
      return nodeFocus.previousElementSibling;
    }
    return _disableListWrap ? null : listNode.lastChild;
  }

  function textCriteriaMatches(nextFocus?: HTMLElement, _textCriteria: typeof textCriteria) {
    if (_textCriteria === undefined) {
      return true;
    }
    let text = nextFocus?.innerText || nextFocus?.textContent;
    text = text.trim().toLowerCase();
    if (text.length === 0) {
      return false;
    }
    if (_textCriteria.repeating) {
      return text[0] === _textCriteria.keys[0];
    }
    return text.indexOf(_textCriteria.keys.join("")) === 0;
  }

  function moveFocus(
    currentFocus?: HTMLElement,
    _disableListWrap?: boolean,
    _disableItemsFocusable?: boolean,
    traversalFunction?: (nodeFocus?: HTMLElement) => null | HTMLInputElement,
    _textCriteria?: typeof textCriteria
  ) {
    let wrappedOnce = false;
    let nextFocus = traversalFunction(currentFocus);

    while (nextFocus) {
      // prevent infinite loop.
      if (nextFocus === listNode.firstChild) {
        if (wrappedOnce) {
          return;
        }
        wrappedOnce = true;
      }

      // same logix as autocomplete.js
      const nextFocusDisabled = _disableItemsFocusable
        ? false
        : nextFocus.disabled || nextFocus.getAttribute("aria-disabled") === "true";

      if (!nextFocus.hasAttribute("tabindex") || !textCriteriaMatches(nextFocus, _textCriteria) || nextFocusDisabled) {
        // move next to element item
        nextFocus = traversalFunction(nextFocus);
      } else {
        nextFocus.focus();
        return;
      }
    }
  }

  // handler
  function handleKeyDown(e: KeyboardEvent) {
    const currentFocus = ownerDocument(listNode).activeElement;
    const key = e.key;

    if (key === "ArrowDown") {
      e.preventDefault();
      moveFocus(currentFocus, disableListWrap, disableItemsFocusable, nextItem);
    } else if (key === "ArrowUp") {
      e.preventDefault();
      moveFocus(currentFocus, disableListWrap, disableItemsFocusable, previousItem);
    } else if ("Home") {
      e.preventDefault();
      moveFocus(null, disableListWrap, disableItemsFocusable, nextItem);
    } else if ("end") {
      e.preventDefault();
      moveFocus(null, disableListWrap, disableItemsFocusable, previousItem);
    } else if (key.length === 1) {
      const _textCriteria = textCriteria;
      const lowerKey = key.toLocaleLowerCase();
      const currTime = performance.now();
      if (_textCriteria.keys.length > 0) {
        // reset
        if (currTime - _textCriteria.lastTime > 500) {
          _textCriteria.keys = [];
          _textCriteria.repeating = true;
          _textCriteria.previousKeyMatched = true;
        } else if (_textCriteria.repeating && lowerKey !== _textCriteria.keys[0]) {
          _textCriteria.repeating = false;
        }
      }
      _textCriteria.lastTime = currTime;
      _textCriteria.keys.push(lowerKey);
      const keepFocusCurrent =
        currentFocus && !_textCriteria.repeating && textCriteriaMatches(currentFocus, _textCriteria);
      if (
        _textCriteria.previousKeyMatched &&
        (keepFocusCurrent || moveFocus(currentFocus, false, disableItemsFocusable, nextItem, _textCriteria))
      ) {
        e.preventDefault();
      } else {
        _textCriteria.previousKeyMatched = false;
      }
    }
    if (onKeyDown) {
      onKeyDown(e);
    }
  }

  // reactive
  $: autoFocus && listNode?.focus();
  $: classes = clsx(className);
</script>

<List {...$$restProps} bind:listNode on:keydown={handleKeyDown} tabindex={autoFocus ? 0 : 1}>
  <slot />
</List>
