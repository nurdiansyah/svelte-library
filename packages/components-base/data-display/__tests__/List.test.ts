import { render, act } from "@testing-library/svelte";
import List from "../List.svelte";
import ListItem from "../list/Item.svelte";
import { writable } from "svelte/store";

describe("List", () => {
  const dataTestId = "list";
  let listProps = {
    "data-testid": dataTestId
  };
  const renderList = (_props = {}) => {
    return render(List, { ...listProps, ..._props }, { container: document.body });
  };
  it("render", () => {
    const { getByTestId, container } = renderList();
    const el = getByTestId(dataTestId);
    expect(el).toBeInTheDocument();
    expect(el).toBeVisible();
    expect(() => el).not.toThrow();
    expect(el).toHaveClass("dbx-list");
    expect(container).toMatchSnapshot();
  });

  it("props: disablePadding = true", async () => {
    let { getByTestId, container, component } = renderList();
    let el = getByTestId(dataTestId);
    expect(el).toHaveClass("dbx-list", "-padding");
    await act(() => {
      component.$set({ disablePadding: true });
    });
    el = getByTestId("list");
    expect(el).not.toHaveClass("-padding");
    expect(container.firstChild).not.toHaveClass("-padding");
  });

  // it("slot: subheader", async () => {
  //   const { getByTestId, container } = render(
  //     h(List, {}, h("header", { slot: "subheader", "data-testid": "slot-list-subheader" }))
  //   );
  //   const el = getByTestId("slot-list-subheader");
  //   expect(el).toBeInTheDocument();
  //   expect(el).toBeVisible();
  //   expect(container).toMatchSnapshot();
  // });
  //
  // it("props: subheader", async () => {
  //   const subheaderProps = h("header", { "data-testid": "props-list-subheader" }, "props-subheader");
  //   const subheader = writable<any>(subheaderProps);
  //   const { getByTestId, container } = renderList({ "bind:subheader": subheader });
  //   let listEl = getByTestId("list");
  //   expect(listEl).toHaveClass("-subheader");
  //   let el = getByTestId("props-list-subheader");
  //   expect(el).toBeInTheDocument();
  //   expect(el).toBeVisible();
  //   expect(container).toMatchSnapshot();
  //   await act(() => {
  //     subheader.set(null);
  //   });
  //   expect(() => getByTestId("props-list-subheader")).toThrow();
  // });
  //
  // it("props: subheader dan slot: subheader, hanya slot subheader yg dipakai", async () => {
  //   const subheaderProps = h("header", { "data-testid": "props-list-subheader" }, "props-subheader");
  //   const subheaderSlot = h("header", { "data-testid": "slot-list-subheader", slot: "subheader" }, "slot-subheader");
  //   const { getByTestId, container } = renderList({ subheader: subheaderProps }, subheaderSlot);
  //   let listEl = getByTestId("list");
  //   expect(listEl).toHaveClass("-subheader");
  //   let el = getByTestId("slot-list-subheader");
  //   expect(el).toBeInTheDocument();
  //   expect(el).toBeVisible();
  //   expect(() => getByTestId("props-list-subheader")).toThrow();
  //   expect(container).toMatchSnapshot();
  // });
  //
  // describe("ListItem", () => {
  //   let listItemProps;
  //   const renderListItem = (_props = listItemProps, context = {}) => {
  //     const fragmentProps = { context: { [KEY_LIST_CONTEXT]: context } };
  //     return render(h(Fragment, fragmentProps, h(ListItem, _props)), {
  //       target: document.body
  //     });
  //   };
  //
  //   beforeEach(() => {
  //     listItemProps = {};
  //   });
  //
  //   it("render", () => {
  //     const { getByTestId, container } = renderListItem();
  //     const el = getByTestId("list-item");
  //     expect(el).toBeInTheDocument();
  //     expect(el).toBeVisible();
  //     expect(() => el).not.toThrow();
  //     expect(el).toHaveClass("list-item");
  //     expect(container).toMatchSnapshot();
  //   });
  //
  //   it("context: { dense: false }, props: { dense: true }", () => {
  //     const { getByTestId } = renderListItem({ dense: true }, { dense: false });
  //     const el = getByTestId("list-item");
  //     expect(el).toHaveClass("-dense");
  //   });
  //
  //   it("context: { dense: false }, props: { dense: false }", () => {
  //     const { getByTestId } = renderListItem({ dense: false }, { dense: false });
  //     const el = getByTestId("list-item");
  //     expect(el).not.toHaveClass("-dense");
  //   });
  // });
});
