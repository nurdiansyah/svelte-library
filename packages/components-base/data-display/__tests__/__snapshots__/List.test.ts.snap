// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List render 1`] = `
<body>
  <div>
    <div
      class="dbx-list -padding"
      data-testid="list"
      role="menu"
    >

    </div>
  </div>
</body>
`;
